File: samples/spaceshooter/scripts/gameplay/enemies.lua
local milk = require "scripts.milk"
local class = require "scripts.class"

local Enemies = class("Enemies")

function Enemies:initialize()
    self.enemies = {
        {xMove = math.random(256), yMove = -20, x = 0, y = 0, sprite = 16, rotation = math.random(20), isAlive = true},
        {xMove = math.random(256), yMove = -20, x = 0, y = 0, sprite = 16, rotation = math.random(20), isAlive = true},
        {xMove = math.random(256), yMove = -40, x = 0, y = 0, sprite = 16, rotation = math.random(20), isAlive = true},
        {xMove = math.random(256), yMove = -80, x = 0, y = 0, sprite = 16, rotation = math.random(20), isAlive = true},
        {xMove = math.random(256), yMove = -100, x = 0, y = 0, sprite = 16, rotation = math.random(20), isAlive = true},
        {xMove = math.random(256), yMove = -120, x = 0, y = 0, sprite = 16, rotation = math.random(20), isAlive = true}
    }
end

function Enemies:getEnemies()
    local liveEnemies = {}
    for i = 1, #self.enemies do
        if self.enemies[i].isAlive then
            table.insert(liveEnemies, self.enemies[i])
        end
    end
    return liveEnemies
end

-- luacheck: push ignore self
function Enemies:destroy(enemy)
    if enemy.isAlive then
        enemy.yMove = -20
        enemy.xMove = math.random(256)
        enemy.rotation = math.random(20)
    end
end
-- luacheck: pop

local function checkOutOfBounds(enemy)
    return enemy.y > 224
end

function Enemies:update()
    for _, enemy in pairs(self:getEnemies()) do
        enemy.yMove = enemy.yMove + 1
        enemy.x = enemy.rotation * math.sin(milk.ticks / 20) + enemy.xMove
        enemy.y = enemy.rotation * math.cos(milk.ticks / 20) + enemy.yMove
        if checkOutOfBounds(enemy) then
            self:destroy(enemy)
        end
    end
end

function Enemies:draw()
    local enemies = self:getEnemies()
    for i = 1, #enemies do
        milk.spr(enemies[i].sprite, enemies[i].x - 8, enemies[i].y - 8)
    end
end

return Enemies

File: sample/scripts/common/enemies.ts
import { ICollidable, CollisionType } from "./collision";
import { Game } from "../game";

const PoolSize = 6;
const EnemySprite = 16;

export interface Enemy extends ICollidable {
    xMove: number;
    yMove: number;
    rotation: number;
    sprite: number;
    health: number;
    color: number;
}

function isOutOfBounds(enemy: Enemy) {
    return enemy.y > 224;
}

export class Enemies {

    private readonly _pool: Enemy[] = [];

    constructor() {
        for (let i = 0; i < PoolSize; i++) {
            this._pool.push({
                type: CollisionType.ENEMY,
                mask: 0,
                x: 0,
                y: 0,
                width: 16,
                height: 16,
                xMove: math.random(256),
                yMove: -math.random(100),
                rotation: math.random(20),
                sprite: EnemySprite,
                health: 2,
                color: 0x00
            });
        }
    }

    get liveEnemies() {
        return this._pool;
    }

    update(game: Game): void {
        for (let i = 0; i < PoolSize; i++) {
            const enemy = this._pool[i];
            enemy.color = 0x00;
            enemy.x = enemy.rotation * math.sin(game.ticks / 20) + enemy.xMove
            enemy.y = enemy.rotation * math.cos(game.ticks / 20) + ++enemy.yMove

            if (isOutOfBounds(enemy)) {
                enemy.health = 2;
                enemy.xMove = math.random(256);
                enemy.yMove = -math.random(100);
                enemy.rotation = math.random(20);
            }
        }
    }

    draw(): void {
        for (let i = 0; i < PoolSize; i++) {
            const enemy = this._pool[i];
            const scale = enemy.y / 224;
            spr(enemy.sprite, enemy.x - 8, enemy.y - 8, 1, 1, scale, 0, enemy.color);
        }
    }

    damage(enemy: Enemy, amount: number): void {
        enemy.health -= amount;
        enemy.color = 0xffffff;
        if (enemy.health <= 0) {
            enemy.health = 2;
            enemy.xMove = math.random(256);
            enemy.yMove = -math.random(100);
            enemy.rotation = math.random(20);
        }
    }
}
